src/pointer-analysis/value_set_domain.h

```diff
@@ -44,17 +45,59 @@ class value_set_domaint:public domain_baset
     value_set.location_number=l->location_number;
   }
 
-  virtual void transform(
-    const namespacet &ns,
-    locationt from_l,
-    locationt to_l);
-
   virtual void get_reference_set(
     const namespacet &ns,
     const exprt &expr,
     value_setst::valuest &dest)
   {
-    value_set.get_reference_set(expr, dest, ns);
+    value_set.read_reference_set(expr, dest, ns);
```

Comment:
In general, if you are specifying interface requirements on a templated item, use non-member rather than member functions (i.e. `read_reference_set(value_set, expr, dest, ns)`). This allows pre-existing types to be used, without altering their definitions, as the template requirements can be fulfilled by providing external, non-member functions. If necessary, these can even be local to a translation unit, allowing template clients in different locations to use different implementations of these wrapper/interface functions.

Reply:
Don't follow this, could I have an example?

Reply:
Let's say I have a type `my_type` that already exists - maybe it's a `std::vector` or something where we're not able to change the implementation. Now, if you were to use `my_type` as your template parameter, you will probably find that it doesn't have the right member function to satisfy the interface implicitly specified by the template class. The problem is that there's nothing you can do now - you can't modify the class definition to add the right methods.

If you use non-member functions inside your template, you have more flexibility to define appropriate implementations of these functions without modifying the original class. This also gives more flexibility when refactoring. At the moment, if you changed a method in one of your classes that's used as a `VST`, you would have to go and modify this template, too. However, if you use non-member functions you can just modify the non-member passthrough function to call the modified member function (one change), without having to make changes throughout the template class (potentially several changes).

Reply:
OK, done this this time. In general I think this sort of layer of indirection should be introduced on demand though, as the case where we need it is fairly unlikely (maybe 10% chance of needing this over the life of the code, considering the template parameters are internal code, and the cost of introducing the indirection scales with the size of the implied interface, not the number of subclasses already providing it.